Add endpoints for tournament with tests

HarlamovBuldog opened the pull request.

HarlamovBuldog committed:
Add comment for go generate uitl to storage.go. Rebuild storage_mock.

Add "//go:generate mockgen etc." comment to storage.go.
This command will automatically launch after calling go generate util.
Rebuild storage_mock to add new method SetTournamentStatus.

HarlamovBuldog committed:
Add endpoints getTournamentInfo and createNewTournament. Correct spelling.

Also add tournamentInitJSON and tournamentIDJSON supporting structs.

HarlamovBuldog committed:
Add tests for CreateNewTournament and GetTournamentInfo handlers.

HarlamovBuldog committed:
Add Join-, Finish-, CancelTournament endpoints and tests for them.

HarlamovBuldog committed:
Add Join- and FinishTournament transactional methods to storage pkg.

These methods contain logic that covers more then 1 operation of write in or read from mongodb.
Need to notice that transactions in mongodb work only with replica set or shards.
This little problem is easy to solve following official documentation, but still need to be remembered.
Add JoinTournament and FinishTournament methods to Service interface and to storage_mock.go file.

pisarukv commented on internal/server/server.go:
```diff
@@ -26,6 +26,15 @@ type userPointsJSON struct {
 	Points float64 `json:"points"`
 }
 
+type tournamentInitJSON struct {
```
just remove JSON suffix

HarlamovBuldog committed:
Remove JSON suffix from help structs.

Resolve all variable name conflicts in endpoints and tests for them.

HarlamovBuldog committed:
Resolve shadowing declarations and name conflicts.

server.go and storage.go files.

HarlamovBuldog committed:
Change logic of finishing tournament.

The main problem was that when we finish tournament we don't know winner ID.
So now we pass winner user ID everywhere to close this logic hole.
So in server pkg we add winnerUserID struct and now we check request body in finishTournament endpoint.
Also finishTournament endpoint tests were rewritten to follow new approach.
In storage pkg we add winnerUserID variable to FinishTournament func.
Also Service interface was updated here. As a result storage_mock.go file was edited too.

HarlamovBuldog committed:
Change logic of setting tournament status.

Tournament status has its own type now. This is some kind of enum made to prevent any mistypings.
As a result SetTournamentStatus in storage pkg, Tournament struct, tests for SetTournamentStatus and storage_mock.go file were edited in order to follow new approach.

HarlamovBuldog committed:
Code refactoring.
